### PR TITLE
[WIP] Active anomaly detection

### PR DESCRIPTION
**Pull request by matwey**

### CODE REVIEW

File: coniferest/calc_paths_sum.pyx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+                    if data[x_index, feature] <= selector.value:
+                        i = selector.left
+                    else:
+                        i = selector.right
+
+                if weights is None:
+                    values[selector.left] += selector.value
+                else:
+                    values[selector.left] += weights[x_index] * selector.value
```
Comment (hombit) [inline coniferest/calc_paths_sum.pyx:116]: `selector.left` is always `-1`

File: coniferest/evaluator.py [hunk 2]
```diff
@@ -52,10 +58,16 @@ def combine_selectors(cls, selectors_list):
...
         indices[1:] = np.add.accumulate(lens)
 
+        # Assign a unique sequential index to every leaf
+        # The index is used for weighted scores
+        leaf_mask = selectors['feature'] < 0
+        leaf_count = np.count_nonzero(leaf_mask)
+        selectors[leaf_mask]['left'] = np.arange(0, leaf_count)
+
         for i in range(len(selectors_list)):
             selectors[indices[i]:indices[i + 1]] = selectors_list[i]
```
Comment (hombit) [inline coniferest/evaluator.py:69]: This should be:
```python
for i in range(len(selectors_list=):
    selectors[indices[i]:indices[i + 1]] = selectors_list[i]

# Assign a unique sequential index to every leaf
# The index is used for weighted scores
leaf_mask = selectors['feature'] < 0
leaf_count = np.count_nonzero(leaf_mask)
selectors['left'][leaf_mask] = np.arange(0, leaf_count)
```
1. `leaf_mask`-related stuff should go after `selectors` initialization
2. `selectors[leaf_mask]['left'] = ` doesn't assign anything

**Comment by hombit** [comment]
Please add some docs to `aadforest.py`

**Comment by hombit** [comment]
Why it is still WIP?

**Comment by balodja** [comment]
It is definitely not WIP. It was already done in march.

**Pull request merged**